app/routes/examples/actions/environment.tsx

```diff
@@ -17,28 +17,29 @@ export const meta: MetaFunction = () => metaTags({ title, description })
 const code = `const schema = z.object({ email: z.string().nonempty().email() })
 
 const environmentSchema = z.object({
-  origin: z.string().regex(/foo.bar/, 'Unauthorized origin'),
+  customHeader: z.string().regex(/foo/, 'Missing custom header'),
 })
 
 const mutation = makeDomainFunction(
   schema,
   environmentSchema,
 )(async (values) => values)
 
-export const action: ActionFunction = async ({ request }) =>
-  formAction({
+export const action: ActionFunction = async ({ request }) => {
+  return formAction({
     request,
     schema,
     mutation,
-    environment: { origin: request.headers.get('Origin') },
+    environment: { customHeader: request.headers.get('customHeader') || '' },
```

Comment:
What should I do with the `null` case in Zod? I put a `|| ''` just to have a working version.

Reply:
Without this, the error is "Expected string got null"

Reply:
I think you could just make it `nullable` on the schema.